src/utils.ts

```diff
@@ -396,3 +398,232 @@ function maybeGetReactionGroupsFallback(
+    for (let i = 0; i < halves.length; i++) {
+      targetMsg = halves[i].find((message) => message.id === id_around);
+      if (targetMsg && i === 0) {
+        hasPrev = false;
+      }
+      if (targetMsg && i === 1) {
+        hasNext = false;
+      }
+    }
+  }
+
+  if (updateHasPrev && typeof hasPrev !== 'undefined') newPagination.hasPrev = hasPrev;
+  if (updateHasNext && typeof hasNext !== 'undefined') newPagination.hasNext = hasNext;
+
+  return newPagination;
+};
+
+const messagePaginationLinear = <StreamChatGenerics extends ExtendableGenerics = DefaultGenerics>({
+  parentSet,
+  requestedPageSize,
+  returnedPage,
+  messagePaginationOptions,
+}: MessagePaginationUpdatedParams<StreamChatGenerics>) => {
+  const newPagination = { ...parentSet.pagination };
+
+  let hasPrev;
+  let hasNext;
+
+  const [firstPageMsg, lastPageMsg] = [returnedPage[0], returnedPage.slice(-1)[0]];
+  const [firstPageMsgIsFirstInSet, lastPageMsgIsLastInSet] = [
+    firstPageMsg?.id && firstPageMsg.id === parentSet.messages[0]?.id,
+    lastPageMsg?.id && lastPageMsg.id === parentSet.messages.slice(-1)[0]?.id,
+  ];
+
+  const queriedNextMessages =
+    messagePaginationOptions &&
+    (messagePaginationOptions.created_at_after_or_equal ||
+      messagePaginationOptions.created_at_after ||
+      messagePaginationOptions.id_gt ||
+      messagePaginationOptions.id_gte);
+
+  const queriedPrevMessages =
+    typeof messagePaginationOptions === 'undefined'
+      ? true
+      : messagePaginationOptions.created_at_before_or_equal ||
+        messagePaginationOptions.created_at_before ||
+        messagePaginationOptions.id_lt ||
+        messagePaginationOptions.id_lte ||
+        messagePaginationOptions.offset;
+
+  const containsUnrecognizedOptionsOnly =
+    !queriedNextMessages &&
+    !queriedPrevMessages &&
+    !messagePaginationOptions?.id_around &&
+    !messagePaginationOptions?.created_at_around;
+
+  const hasMore = returnedPage.length >= requestedPageSize;
+
+  if (typeof queriedPrevMessages !== 'undefined' || containsUnrecognizedOptionsOnly) {
+    hasPrev = hasMore;
+  }
+  if (typeof queriedNextMessages !== 'undefined') {
+    hasNext = hasMore;
+  }
+  const returnedPageIsEmpty = returnedPage.length === 0;
+
+  if ((firstPageMsgIsFirstInSet || returnedPageIsEmpty) && typeof hasPrev !== 'undefined')
+    newPagination.hasPrev = hasPrev;
+  if ((lastPageMsgIsLastInSet || returnedPageIsEmpty) && typeof hasNext !== 'undefined')
+    newPagination.hasNext = hasNext;
+
+  return newPagination;
+};
+
+export const messageSetPagination = <StreamChatGenerics extends ExtendableGenerics = DefaultGenerics>(
+  params: MessagePaginationUpdatedParams<StreamChatGenerics>,
+) => {
+  if (params.parentSet.messages.length < params.returnedPage.length) {
+    console.error('Corrupted message set state: parent set size < returned page size');
+    return params.parentSet.pagination;
+  }
+
+  if (params.messagePaginationOptions?.created_at_around) {
```

Comment:
If the query options object contains multiple params, then the pagination is determined as if there was only one of them in the following order: 1. created_at_around, 2. id_around and 3. the rest of the query params.